src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+                }
+            }
+        };
+    };
+
+    /**
+     * Given a shard that has reads and variants, query Google Genomics' Reference server and get reference info,
+     * and fill that and the correct variants into readContext.
+     */
+    public static DoFn<ContextShard, ContextShard> fillContext(final ReferenceDataflowSource refSource, final PCollectionView<SAMSequenceDictionary> refDictView, final GenomicsFactory.OfflineAuth auth) throws IOException {
+        return new DoFnWLog<ContextShard, ContextShard>("fillContext") {
+            private static final long serialVersionUID = 1L;
+            private transient Stopwatch getRef, queryVariants, sending, total;
+            private int elementCount;
+
+            @Override
+            public void startBundle(Context c) throws Exception {
+                super.startBundle(c);
+                getRef = Stopwatch.createUnstarted();
+                queryVariants = Stopwatch.createUnstarted();
+                sending = Stopwatch.createUnstarted();
+                total = Stopwatch.createUnstarted();
+                elementCount=0;
+            }
+
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                total.start();
+                elementCount++;
+                ContextShard shard = c.element();
+                SAMSequenceDictionary refDict = c.sideInput(refDictView);
+                int margin = 500;
+                int start = shard.interval.getStart();
+                if (start>margin) start -= margin;
+                else start = 1; // it's 1-based, remember.
+                int end = shard.interval.getEnd();
+                getRef.start();
+                int len = refDict.getSequence(shard.interval.getContig()).getSequenceLength();
+                getRef.stop();
+                if (end<len-margin) end+= margin;
+                else end = len; // OK because those are 1-based closed intervals
+                SimpleInterval refInterval = new SimpleInterval(shard.interval.getContig(), start, end);
+                getRef.start();
+                ReferenceBases refBases = refSource.getReferenceBases(c.getPipelineOptions(), refInterval);
+                getRef.stop();
+
+                ArrayList<ReadContextData> readContext = new ArrayList<>();
+                for (GATKRead r : shard.reads) {
+                    SimpleInterval readInterval = new SimpleInterval(r.getContig(), r.getStart(), r.getEnd());
+                    queryVariants.start();
+                    ArrayList<Variant> variantsOverlappingThisRead = shard.variantsOverlapping(readInterval);
+                    queryVariants.stop();
+                    // we pass all the bases. That's better because this way it's just a shared
+                    // pointer instead of being an array copy. Downstream processing is fine with having
+                    // extra bases (it expects a few, actually).
```

Comment:
Again, see my note about `BQSR_REFERENCE_WINDOW_FUNCTION` above -- you might in theory not have all the bases you need here...